### PR TITLE
[#IP-431] add function_app module

### PR DESCRIPTION
**Pull request by fabriziopapi**
Add function_app module configuration

### CODE REVIEW

**Review by uolter — COMMENTED** [review]
Not blocking 
https://www.terraform-best-practices.com/naming
> Do not repeat resource type in resource name (not partially, nor completely):
Good: resource "aws_route_table" "public" {}
Bad: resource "aws_route_table" "public_route_table" {}
Bad: resource "aws_route_table" "public_aws_route_table" {}

> **Resource name should be named this if there is no more descriptive and general name** available, or if resource module creates single resource of this type (eg, there is single resource of type aws_nat_gateway and multiple resources of typeaws_route_table, so aws_nat_gateway should be named this and aws_route_table should have more descriptive names - like private, public, database).

File: function_app/main.tf [hunk 1]
```diff
@@ -0,0 +1,278 @@
+data "azurerm_key_vault_secret" "allowed_ips_secret" {
+  count = var.allowed_ips_secret == null ? 0 : 1
+
+  name         = var.allowed_ips_secret.key_vault_secret
+  key_vault_id = var.allowed_ips_secret.key_vault_id
+}
```
Comment (pasqualedevita) [inline function_app/main.tf:6]:
```suggestion
```
u can delete this, now we can pass these values in safe mode without this

File: function_app/main.tf [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+  tags = var.tags
+}
+
+module "secrets_from_keyvault" {
+  source = "git::https://github.com/pagopa/azurerm.git//key_vault_secrets_query?ref=v1.0.58"
+
+  key_vault_name    = var.app_settings_secrets.name
+  resource_group    = var.app_settings_secrets.resource_group
+  secrets           = var.app_settings_secrets.secrets
+}
```
Comment (pasqualedevita) [inline function_app/main.tf:163]:
```suggestion
```
u can delete this, now we can pass these values in safe mode without this

File: function_app/main.tf [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+      SLOT_TASK_HUBNAME        = "ProductionTaskHub"
+      WEBSITE_RUN_FROM_PACKAGE = 1
+      WEBSITE_VNET_ROUTE_ALL   = 1
+      WEBSITE_DNS_SERVER       = "168.63.129.16"
+      # this app settings is required to solve the issue:
+      # https://github.com/terraform-providers/terraform-provider-azurerm/issues/10499
+      WEBSITE_CONTENTSHARE = "${local.resource_name}-content"
+    },
+    var.app_settings,
+    module.secrets_from_keyvault.secrets_with_value,
```
Comment (pasqualedevita) [inline function_app/main.tf:227]:
```suggestion
```
same

File: function_app/main.tf [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    service_delegation = {
+      name    = "Microsoft.Web/serverFarms"
+      actions = ["Microsoft.Network/virtualNetworks/subnets/action"]
+    }
+  }
+
+  service_endpoints = [
+    "Microsoft.Web"
+  ]
+}
```
Comment (pasqualedevita) [inline function_app/main.tf:271]:
```suggestion
```
u can delete this, we create subnet outside fnapp

File: function_app/main.tf [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+
+resource "azurerm_app_service_virtual_network_swift_connection" "app_service_virtual_network_swift_connection" {
+  count = var.subnet_id == null && var.virtual_network_info == null ? 0 : 1
+
+  app_service_id = azurerm_function_app.function_app.id
+  subnet_id      = local.subnet_id
+}
```
Comment (pasqualedevita) [inline function_app/main.tf:278]: this is important do not delete it :)

**Review by pasqualedevita — CHANGES_REQUESTED** [review]
u can try a plan creating a new branch on `io-infra` and pointing ref to this branch

File: function_app/main.tf (hunk 5 above)
Comment (fabriziopapi) [reply function_app/main.tf:242]: I will keep it safe and sound!

**Comment by fabriziopapi** [comment]
> Not blocking https://www.terraform-best-practices.com/naming
> 
> > Do not repeat resource type in resource name (not partially, nor completely):
> > Good: resource "aws_route_table" "public" {}
> > Bad: resource "aws_route_table" "public_route_table" {}
> > Bad: resource "aws_route_table" "public_aws_route_table" {}
> 
> > **Resource name should be named this if there is no more descriptive and general name** available, or if resource module creates single resource of this type (eg, there is single resource of type aws_nat_gateway and multiple resources of typeaws_route_table, so aws_nat_gateway should be named this and aws_route_table should have more descriptive names - like private, public, database).

Fixed!

**Review by uolter — APPROVED** [review]

File: function_app/variables.tf [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+variable "subnet_id" {
+  type = string
+}
+
+variable "virtual_network_info" {
+  type = object({
+    name                  = string
+    resource_group_name   = string
+    subnet_address_prefix = string
+  })
```
Comment (pasqualedevita) [inline function_app/variables.tf:128]:
```suggestion
```

File: function_app/main.tf [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+  count = var.export_keys ? 1 : 0
+
+  name                = local.resource_name
+  resource_group_name = var.resource_group_name
+  depends_on          = [azurerm_function_app.this]
+}
+
+
+resource "azurerm_app_service_virtual_network_swift_connection" "this" {
+  count = var.virtual_network_info == null ? 0 : 1
```
Comment (pasqualedevita) [inline function_app/main.tf:240]:
```suggestion
  count = var. subnet_id == null ? 0 : 1
```

File: function_app/variables.tf (hunk 6 above)
Comment (pasqualedevita) [reply function_app/variables.tf:128]: remove `virtual_network_info`

File: function_app/variables.tf [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+variable "avoid_old_subnet_delete" {
+  type = bool
+
+  default = false
+}
+
+# variable "export_keys" {
+#   type    = bool
+#   default = false
+# }
```
Comment (pasqualedevita) [inline function_app/variables.tf:160]:
```suggestion
```
not used

File: function_app/variables.tf [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+  default = 10
+}
+
+# variable "web_test" {
+#   type = object({
+#     application_insights_id = string
+#     enabled                 = bool
+#   })
+#   default = null
+# }
```
Comment (pasqualedevita) [inline function_app/variables.tf:178]:
```suggestion
```
not used

File: function_app/variables.tf [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+
+variable "advanced_threat_protection_enable" {
+  type        = bool
+  default     = false
+  description = "Enable advanced threat protection on the storage accounts"
+}
```
Comment (pasqualedevita) [inline function_app/variables.tf:184]:
```suggestion
```
not used

**Pull request merged**